core/common/src/main/kotlin/tachiyomi/core/common/archive/ArchiveReader.kt

```diff
@@ -0,0 +1,42 @@
+package tachiyomi.core.common.archive
+
+import android.content.Context
+import android.os.ParcelFileDescriptor
+import android.system.Os
+import android.system.OsConstants
+import com.hippo.unifile.UniFile
+import me.zhanghai.android.libarchive.ArchiveException
+import tachiyomi.core.common.storage.openFileDescriptor
+import java.io.Closeable
+import java.io.InputStream
+
+class ArchiveReader(pfd: ParcelFileDescriptor) : Closeable {
+    val size = pfd.statSize
+    val address = Os.mmap(0, size, OsConstants.PROT_READ, OsConstants.MAP_PRIVATE, pfd.fileDescriptor, 0)
+
+    inline fun <T> useEntries(block: (Sequence<ArchiveEntry>) -> T): T =
+        ArchiveInputStream(address, size).use { block(generateSequence { it.getNextEntry() }) }
+
+    fun getInputStream(entryName: String): InputStream? {
+        val archive = ArchiveInputStream(address, size)
+        try {
+            while (true) {
+                val entry = archive.getNextEntry() ?: break
+                if (entry.name == entryName) {
+                    return archive
+                }
+            }
+        } catch (e: ArchiveException) {
+            archive.close()
+            throw e
+        }
+        archive.close()
```

Comment:
use `finally`?

Reply:
`archive` should not be closed if it's returned.